src/transport/PeerMessageCounter.h

```diff
@@ -0,0 +1,159 @@
+    bool IsSynchronizing() { return mStatus == Status::SyncInProcess; }
+    bool IsSyncCompleted() { return mStatus == Status::Synced; }
+
+    void StartSync(std::array<uint8_t, kChallengeSize> challenge)
+    {
+        mStatus                   = Status::SyncInProcess;
+        mSyncInProcess.mChallenge = challenge;
+    }
+
+    void SyncFail() { Reset(); }
+
+    CHIP_ERROR VerifyChallenge(uint32_t counter, std::array<uint8_t, kChallengeSize> challenge)
+    {
+        if (mStatus != Status::SyncInProcess)
+            return CHIP_ERROR_INCORRECT_STATE;
+        if (mSyncInProcess.mChallenge != challenge)
+            return CHIP_ERROR_INVALID_ARGUMENT;
+
+        mStatus             = Status::Synced;
+        mSynced.mMaxCounter = counter;
+        mSynced.mWindow.set();         // set all bits, deny all packets with counter less than the given counter
+        mSynced.mWindow.set(0, false); // expect next packet with counter equal to the given counter
+        return CHIP_NO_ERROR;
+    }
+
+    CHIP_ERROR Verify(uint32_t counter) const
+    {
+        if (mStatus != Status::Synced)
+            return CHIP_ERROR_INCORRECT_STATE;
+
+        if (counter <= mSynced.mMaxCounter)
+        {
+            uint32_t offset = mSynced.mMaxCounter - counter;
+            if (offset >= CHIP_CONFIG_MESSAGE_COUNTER_WINDOW_SIZE)
+                return CHIP_ERROR_INVALID_ARGUMENT; // outside valid range
+            if (mSynced.mWindow.test(offset))
+                return CHIP_ERROR_INVALID_ARGUMENT; // duplicated, in window
+        }
+
+        return CHIP_NO_ERROR;
+    }
+
+    /**
+     * @brief
+     *    With the counter verified and the packet MAC is also verified by the secure key, we can trust the packet and adjust
+     *    counter states.
+     *
+     * @pre Verify(counter) == CHIP_NO_ERROR
+     */
+    void Commit(uint32_t counter)
+    {
+        if (counter <= mSynced.mMaxCounter)
+        {
+            uint32_t offset = mSynced.mMaxCounter - counter;
+            mSynced.mWindow.set(offset);
+        }
+        else
+        {
+            uint32_t offset = counter - mSynced.mMaxCounter;
+            // advance max counter by `offset`
+            mSynced.mMaxCounter = counter;
+            if (offset < CHIP_CONFIG_MESSAGE_COUNTER_WINDOW_SIZE)
+            {
+                mSynced.mWindow <<= offset;
+            }
+            else
+            {
+                mSynced.mWindow.reset();
+            }
+            mSynced.mWindow.set(0, true);
+        }
+    }
+
+    /* Test-only */
+    void SetCounter(uint32_t value)
+    {
+        Reset();
+        mStatus             = Status::Synced;
+        mSynced.mMaxCounter = value;
+        mSynced.mWindow.set();
+        mSynced.mWindow.set(0, false);
+    }
+
+    /* Test-only */
+    uint32_t GetCounter() { return mSynced.mMaxCounter; }
+
+private:
+    enum class Status
+    {
+        NotSync,
+        SyncInProcess,
+        Synced,
+    } mStatus;
+
+    union
```

Comment:
Please document which branch of the union is expected to be active when (depending on mStatus).

Reply:
I think the member name is already giving enough information about it.

Reply:
I asked for the documentation because it wasn't immediately clear to me, actually.  The fact that there are three states but only two union branches did not help with that, for example.

Reply:
Added comment in the enum class